diff -aur hal-0.5.8.vanilla/hald/linux/addons/addon-acpi.c hal-0.5.8/hald/linux/addons/addon-acpi.c
--- hal-0.5.8.vanilla/hald/linux/addons/addon-acpi.c	2006-09-11 23:56:39.000000000 +0200
+++ hal-0.5.8/hald/linux/addons/addon-acpi.c	2006-09-17 16:12:51.000000000 +0200
@@ -32,6 +32,7 @@
 #include <string.h>
 #include <sys/socket.h>
 #include <sys/types.h>
+#include <sys/stat.h>
 #include <sys/un.h>
 #include <unistd.h>
 
@@ -45,9 +46,14 @@
 acpi_get_event_fp_kernel (void)
 {
 	FILE *fp = NULL;
+	struct stat sbuf;
 
-	fp = fopen ("/proc/acpi/event", "r");
+	if (stat("/usr/sbin/acpid", &sbuf) == 0)  {
+		HAL_DEBUG (("acpid installed, not using the kernel acpi event interface"));
+		return NULL;
+	}
 
+	fp = fopen ("/proc/acpi/event", "r");
 	if (fp == NULL)
 		HAL_ERROR (("Cannot open /proc/acpi/event: %s", strerror (errno)));
 
Only in hal-0.5.8/hald/linux/addons: addon-acpi.c.orig
